src/Token/TokenProvider.php

```diff
@@ -0,0 +1,269 @@
+     * a token.
+     *
+     * @return callable
+     */
+    public static function chain()
+    {
+        $links = func_get_args();
+        //Common use case for when aws_shared_config_files is false
+        if (empty($links)) {
+            return function () {
+                return Promise\Create::promiseFor(false);
+            };
+        }
+
+        return function () use ($links) {
+            /** @var callable $parent */
+            $parent = array_shift($links);
+            $promise = $parent();
+            while ($next = array_shift($links)) {
+                $promise = $promise->otherwise($next);
+            }
+            return $promise;
+        };
+    }
+
+    /**
+     * Wraps a token provider and caches a previously provided token.
+     * Ensures that cached tokens are refreshed when they expire.
+     *
+     * @param callable $provider Token provider function to wrap.
+     * @return callable
+     */
+    public static function memoize(callable $provider)
+    {
+        return function () use ($provider) {
+            static $result;
+            static $isConstant;
+
+            // Constant tokens will be returned constantly.
+            if ($isConstant) {
+                return $result;
+            }
+
+            // Create the initial promise that will be used as the cached value
+            // until it expires.
+            if (null === $result) {
+                $result = $provider();
+            }
+
+            // Return a token that could expire and refresh when needed.
+            return $result
+                ->then(function (TokenInterface $token) use ($provider, &$isConstant, &$result) {
+                    // Determine if the token is constant.
+                    if (!$token->getExpiration()) {
+                        $isConstant = true;
+                        return $token;
+                    }
+
+                    if (!$token->isExpired()) {
+                        return $token;
+                    }
+                    return $result = $provider();
+                })
+                ->otherwise(function($reason) use (&$result) {
+                    // Cleanup rejected promise.
+                    $result = null;
+                    return Promise\Create::promiseFor(null);
+                });
+        };
+    }
+
+    /**
+     * Wraps a token provider and saves provided token in an
+     * instance of Aws\CacheInterface. Forwards calls when no token found
+     * in cache and updates cache with the results.
+     *
+     * @param callable $provider Token provider function to wrap
+     * @param CacheInterface $cache Cache to store the token
+     * @param string|null $cacheKey (optional) Cache key to use
+     *
+     * @return callable
+     */
+    public static function cache(
+        callable $provider,
+        CacheInterface $cache,
+        $cacheKey = null
+    ) {
+        $cacheKey = $cacheKey ?: 'aws_cached_token';
```

Comment:
What happens if this is an empty string?

Reply:
An empty string is falsy, a string with a space in it though would use the space as the cache key.